## Chapter Six Results

## <font color='green'>Application: Make the initial guess 7% and calculate two new guesses for the following bonds:</font>

<div style="
    border-left: 12px solid green;
    line-height: 1.5;
    padding: 15px">
<br>



|Maturity|Coupon|Spot Rate|
|-------|-------|---------|
|&nbsp;&nbsp;&nbsp;Six Months|&nbsp;&nbsp;&nbsp;0|&emsp;4%|
|&nbsp;&nbsp;&nbsp;One Year|&nbsp;&nbsp;&nbsp;0|&emsp;5%|
|&nbsp;&nbsp;&nbsp;One Year|&nbsp;&nbsp;&nbsp;5|&emsp;|
<br>




1. **Calculate bond prices at spot rates**  
   * Import Numpy for continuous compounding function np.exp()
   * Calculate present value factors: 0.5 years and 1.0 years at spot rates  
   * Bond prices equal sum of cash flow present values
2. **Calculate bond prices and first derivatives at first guess rate = 7%**  
3. **Update prices and values with two more guesses**  

</div>

In [1]:
try:
    import numpy as np
except:
    !pip install numpy
    import numpy as np

In [2]:
# actual spot rates = .04, .05
# present value factors
pv_6mo= np.exp(-.04/2)
pv_12mo = np.exp(-.05)
print(pv_6mo ,pv_12mo )

0.9801986733067553 0.951229424500714


In [3]:
# bond cash flows intialized as tuples
cf_6mo_zero=(100,0)
cf_12mo_zero=(0,100)
cf_12mo_coupon5=(2.5,102.5)
# bond value equals sum of cash flow present values
bond_value_6mo_zero=pv_6mo*cf_6mo_zero[0]+pv_12mo*cf_6mo_zero[1]
bond_value_12mo_zero=pv_6mo*cf_12mo_zero[0]+pv_12mo*cf_12mo_zero[1]
bond_value_12mo_coupon5=pv_6mo*cf_12mo_coupon5[0]+pv_12mo*cf_12mo_coupon5[1]
bond_prices=np.array([bond_value_6mo_zero,bond_value_12mo_zero,bond_value_12mo_coupon5])
print('Bond Prices = ',bond_prices)

Bond Prices =  [98.01986733 95.12294245 99.95151269]


In [4]:
# solving for yield to maturity
# initial rate guess for each bond = 7%
guess=np.array([.07,.07,.07])
# subsequent guesses need to remove comment sign from next statement
guess=guess_new
# present value factors
pv_6mo = np.exp(-guess[0]/2)
pv_12mo = np.exp(-guess[1])
pv_coupon = np.exp(-guess[2])

bond_value_6mo_zero=pv_6mo*cf_6mo_zero[0]+pv_6mo*cf_6mo_zero[1]
bond_value_12mo_zero=pv_12mo*cf_12mo_zero[0]+pv_12mo*cf_12mo_zero[1]
bond_value_12mo_coupon5=pv_coupon*cf_12mo_coupon5[0]+pv_coupon*cf_12mo_coupon5[1]
bond_values=np.array([bond_value_6mo_zero,bond_value_12mo_zero,bond_value_12mo_coupon5])
print('Bond Value Guesses = ',bond_values)

NameError: name 'guess_new' is not defined

In [ ]:
# price errors equal actual bond price - guess value
errors=bond_prices-bond_values
errors

array([-0.01108321, -0.01915398, -0.04684359])

In [ ]:
# first derivative at guess rate
# first derivative equals negative  sum of cash flow present values
# weighted by year (0.5 or 1.0)

derivative_6mo_zero=    -(pv_6mo*cf_6mo_zero[0]*.5 +    pv_12mo*cf_6mo_zero[1]*1.0)
derivative_12mo_zero=   -(pv_12mo*cf_12mo_zero[0]*0.5 +   pv_12mo*cf_12mo_zero[1])
derivative_12mo_coupon5=-(pv_coupon*cf_12mo_coupon5[0]*0.5 + pv_coupon*cf_12mo_coupon5[1]*1.0)
bond_derivatives=np.array([derivative_6mo_zero,derivative_12mo_zero,derivative_12mo_coupon5])
print('Derivative Values = ',bond_derivatives)

Derivative Values =  [-49.01547527 -95.14209643 -98.80789966]


In [ ]:
# errors are scaled by derivative values to find increment to guess
delta_guess=errors/bond_derivatives
delta_guess

array([0.00022612, 0.00020132, 0.00047409])

In [ ]:
guess_new=guess+delta_guess
guess_new

array([0.03999999, 0.04999998, 0.04928069])

Go back to "solving for yield to maturity" code cell and remove comment sign (#) from:
guess=guess_new
Rerun code cells to see new bond values, errors, derivatives and new rate guesses

   **First guess of 7% gives:**  
   * delta_guess = array([-0.03022613, -0.02020134, -0.0211934 ])  
   * guess-new =   array([0.03977387, 0.04979866, 0.0488066 ])

   **Second guess of [0.03977387, 0.04979866, 0.0488066 ] gives:**  
   * delta_guess = array([0.00022612, 0.00020132, 0.00047409]))  
   * guess-new =   array([0.03999999, 0.04999998, 0.04928069])   

   Yields to maturity are discovered after two rounds of
   4.0%, 5.0% and 4.92% for the three bonds.


## <font color='green'>Application: For January 31$^{st}$ 2030 example update guess from 5% and recalculate present value with updated guess</font>

<br>
Use the present value calculated with spot rates as the target value.

<br>


<div style="
    border-left: 12px solid green;
    line-height: 1.5;
    padding: 15px">
<br>


1. **Import Required Modules:**  
   * Built-in modules  
   * Pandas (install if necessary)  
   * Custom modules:  
     * <font color='green'>accrued_interest</font>  
     * <font color='green'>create_payoff_matrix</font>  
     * <font color='green'>bond_pay_data</font>  
2. **Add user defined functions:**  
   *  <font color='green'>bond_pv</font>  
3. **Download Data:**  
   * Retrieve the <font color='green'>bond_data</font> file from DropBox.  
4. **Perform Calculations:**  
   * Using 5% as first guess. execute <font color='green'>bond_pv</font> for present value and derivative values
   * Update price and values with one more guess   

</div>

### Required modules

In [ ]:
# Import OS to interact with local computer operating system
import sys
import requests
from types import ModuleType
# Import the datetime and date classes from the datetime module
from datetime import datetime, date

# Import the Pandas library for data manipulation and analysis, aliased as pd.
try:
    import pandas as pd
except:
    !pip install pandas
    import pandas as pd

# Import the NumPy library for data manipulation and analysis, aliased as np.
try:
    import numpy as np
except:
    !pip install numpy
    import numpy as np

In [ ]:
# Define the URL of the Python module to be downloaded from Dropbox.
# The 'dl=1' parameter in the URL forces a direct download of the file content.
url= 'https://www.dropbox.com/scl/fi/4y5hjxlfphh1ngvbgo77q/\
module_-basic_concepts_fixed_income.py?rlkey=6oxi7mgka42veaat79hcv8boz&st=87sztshr&dl=1'
module_name='basic_concepts_fixed_income'
# Send an HTTP GET request to the URL and store the server's response.
try:
  response=requests.get(url)
  # Raise an exception for bad status codes (like 404 Not Found)
  response.raise_for_status()
  module= ModuleType(module_name)
  #Code contained in response.text executed
  exec(response.text, module.__dict__)
  # Module added to sys
  sys.modules[module_name]=module
except requests.exceptions.RequestException as e:
    print(f"❌ Error: Could not fetch module from URL. {e}")
except Exception as e:
    print(f"❌ Error: Failed to execute or import the module. {e}")

# Now that 'basic_concepts_fixed_income' exists in the notebook, import the specific functions
from basic_concepts_fixed_income import (accrued_interest,
                                         create_payoff_matrix,
                                         bond_pay_data)

### Calculating present value and derivatives: the <font color='green'>bond_pv</font> function


In [ ]:
def bond_pv(rates=None, data_dict=None, settlement=None):
  '''Calculates a bond's present value, and its first and second derivatives.

  This function prices a bond by discounting its future cash flows using a
  provided set of continuously compounded interest rates. It also computes the
  first and second derivatives of the present value with respect to the rates,
  which are fundamental inputs for calculating risk measures like dollar duration
  and convexity.

  The time to payment is calculated using an Actual/365.25 day-count convention.
  If the provided array of rates is shorter than the number of payments, the
  last rate is used to extrapolate for all subsequent payments.

  Args:
      rates (float or np.ndarray): The continuously compounded discount rate or an
          array of rates. If a single float is provided, it's applied to all
          cash flows.
      data_dict (dict): A dictionary that must contain the bond's payment data.
          It can optionally specify the settlement date.
          - 'pay_data' (tuple): A required tuple of two NumPy arrays:
            (payment_dates, payment_amounts).
          - 'settlement' (datetime.date): An optional valuation date.
      settlement (datetime.date, optional): The valuation date. This argument is
          used if a settlement date is not found in `data_dict`. It defaults
          to the current system date if not provided elsewhere.

  Returns:
      tuple: A tuple of three floats:
      - Present Value: The bond's price,.
      - First Derivative: The derivative of PV with respect to rates. This is related
        to the bond's dollar duration.
      - Second Derivative: The second derivative of PV with respect to rates.
        This is related to the bond's dollar convexity.

  Raises:
      TypeError: If 'pay_data' is not a tuple of two NumPy arrays or if
                 'rates' is not numeric.
      ValueError: If the dates and payments arrays within 'pay_data' are not
                  the same size.
  '''
  # Import necessary libraries
  import numpy as np
  from datetime import date

  # if settlement in dictionary otherwise value passed is no settlement None
  settlement=data_dict.get('settlement',settlement)
  # settlement date is today if it's not provided
  if settlement is None:
    settlement=date.today()
  pay_data=data_dict['pay_data']

# Validate the structure and types within pay_data
  if not isinstance(pay_data, tuple) or len(pay_data) != 2:
      raise TypeError("'pay_data' must be a tuple of (dates_array, payments_array).")
  if not (isinstance(pay_data[0], np.ndarray) and isinstance(pay_data[1], np.ndarray)):
      raise TypeError("Both items in 'pay_data' must be NumPy arrays.")
  if pay_data[0].size != pay_data[1].size:
      raise ValueError("Dates and payments arrays in 'pay_data' must have the same size.")
  # --- Input Validation ---
    # Convert rates to a NumPy array and validate
  if not hasattr(rates, '__iter__'):
      rates = np.array([rates])
  else:
      rates = np.array(rates)

  if not np.issubdtype(rates.dtype, np.number):
        raise TypeError("'rates' must contain only numeric data.")

  # Calculate the time to each payment in years from the settlement date
  # Convert dates to NumPy datetime64, float difference convert to years
  pay_dates=(np.array(pay_data[0],dtype='datetime64[D]')
             -np.datetime64(settlement)).astype(float)/365.25

  # Ensure there is a discount rate for every payment date
  if rates.size<pay_dates.size:
  #If single rate create an array of that single rate
    if rates.size==1:
      rates=np.full(pay_dates.size,rates[0])
   #If more than one but less then size of pay dates
    else:
      rates=np.append(rates,np.full(pay_dates.size-rates.size,rates[-1]))

  #Calculate pv of each payment and sum to get value
  pv_payments=pay_data[1]*np.exp(-rates*pay_dates)
  value=np.sum(pv_payments)

  #Derivatives for each payment
  derivatives_array=-pay_dates*pv_payments

  #First derivative
  first_derivative=np.sum(derivatives_array)

  #Second derivative
  second_derivative=np.sum(-pay_dates*derivatives_array)
  return value,first_derivative,second_derivative

### Download bond data

In [ ]:
#The full file path.
url='https://www.dropbox.com/scl/fi/lgnaj41bt8o9sv5a63rr1/\
bond_data_jan21_2025.xlsx?rlkey=twjzkcqo0g2ahvot78518ti4x&st=ihc5feog&dl=1'
print(f"Attempting to load data from: {url}")

#Load the data from Excel, using the first column as the index.
try:
    bond_data = pd.read_excel(url, index_col='Maturity Date',sheet_name='Fidelity Data')

    # Display the first and last 5 rows of the loaded DataFrame to verify it worked.
    display(bond_data)

except FileNotFoundError:
    print("\nERROR: File not found.")
    print("Please check that the 'folder' and 'file' variables are spelled correctly'\
' and that the file exists in that location.")

Attempting to load data from: https://www.dropbox.com/scl/fi/lgnaj41bt8o9sv5a63rr1/bond_data_jan21_2025.xlsx?rlkey=twjzkcqo0g2ahvot78518ti4x&st=ihc5feog&dl=1


,Description,Coupon,Price Bid,Price Ask,Bid Size,Ask Size
Maturity Date,,,,,,
2025-01-28,UNITED STATES TREAS BILLS ZERO CPN 0.000...,0.000,99.929,99.930,100000,100000
2025-01-30,UNITED STATES TREAS BILLS ZERO CPN 0.000...,0.000,99.906,99.907,40000,40000
2025-01-31,UNITED STATES TREAS SER U-2025 1.3750...,1.375,99.921,99.934,60000,60000
2025-01-31,UNITED STATES TREAS SER AW-2025 4.1250...,4.125,99.988,99.997,60000,60000
2025-01-31,UNITED STATES TREAS SER G-2025 2.5000...,2.500,99.953,99.965,60000,60000
...,...,...,...,...,...,...
2029-11-30,UNITED STATES TREAS SER AG-2029 4.1250...,4.125,98.910,98.914,100000,100000
2029-11-30,UNITED STATES TREAS SER S-2029 3.8750...,3.875,97.763,97.782,40000,40000
2029-12-31,UNITED STATES TREAS SER T-2029 3.8750...,3.875,97.734,97.738,65000,65000


## Estimating the spot rates for the 149 dates

In [ ]:
settlement=date(2025,1,21)
# apply the accrued_interest function to the DataFrame
bond_data['Accrued'] = bond_data.apply(
    lambda x: accrued_interest(
        maturity=x.name, # Assuming index is maturity
        coupon=x['Coupon'],
        settlement=settlement,
        freq=2,
        day_type='Actual/Actual'
    ), axis=1
)

# Calculate transaction prices
transaction_prices=((bond_data['Price Bid']+bond_data['Price Ask'])/2+
                    bond_data['Accrued'])

# Create payoff matrix
payoffs,column_dates=create_payoff_matrix(bond_data,settlement=settlement,freq=2)

# Estimate the present value factors with least squares function of linalg module
zero_prices,ss2,rank,colInfo=np.linalg.lstsq(payoffs,transaction_prices)

# Vectorize calculation of maturity in years with NumPy
# Maturity of the bonds in years..convert date object to NumPy datetime64
maturity_years=(np.array(column_dates,dtype='datetime64[D]')-
                np.datetime64(settlement)).astype(float)/365.25

# Continuously compounded annualized spot rates
# name assigned string 'Spot Rates'
spot_rates=-pd.Series(np.log(zero_prices)/np.array(maturity_years),
                      index=column_dates,name='Spot Rates')
display(spot_rates)

,Spot Rates
2025-01-28,0.036799
2025-01-30,0.037963
2025-01-31,0.040711
2025-02-04,0.039817
2025-02-06,0.040213
...,...
2029-10-31,0.043390
2029-11-15,0.043874
2029-11-30,0.043402
2029-12-31,0.043400


In [ ]:
# Get characteristics from bond_data DataFrame (last row of DataFrame)
ustreasury_2030_1_31=bond_data.iloc[-1]

# Assign values
# maturity last value of index, coupon from  column name, settlement 1/21/2025
maturity=bond_data.index[-1]
coupon= ustreasury_2030_1_31['Coupon']
settlement=date(2025,1,21)

# Get pay dates and amounts for the bond
pay_data=bond_pay_data(maturity,coupon,settlement=settlement,freq=2)

# data dictionary
data_dict={'settlement':settlement,'pay_data':pay_data}

data_dict

{'settlement': datetime.date(2025, 1, 21),
 'pay_data': (array([datetime.date(2025, 1, 31), datetime.date(2025, 7, 31),
         datetime.date(2026, 2, 2), datetime.date(2026, 7, 31),
         datetime.date(2027, 2, 1), datetime.date(2027, 8, 2),
         datetime.date(2028, 1, 31), datetime.date(2028, 7, 31),
         datetime.date(2029, 1, 31), datetime.date(2029, 7, 31),
         datetime.date(2030, 1, 31)], dtype=object),
  array([  1.75,   1.75,   1.75,   1.75,   1.75,   1.75,   1.75,   1.75,
           1.75,   1.75, 101.75]))}

In [ ]:
# results for yield to maturity
guess=.05
ytm_5=bond_pv(guess,data_dict=data_dict)
# results from estimated spot rates
term_structure=bond_pv(spot_rates.loc[pay_data[0]],data_dict=data_dict)

# create DataFrame of results
results_dict={'5% Yield To Maturity':ytm_5,'Spot Rates':term_structure}
display(pd.DataFrame(results_dict,index=['Values','1st Derivatives','2nd Derivatives']))

,5% Yield To Maturity,Spot Rates
Values,94.795194,97.657391
1st Derivatives,-431.824569,-445.720259
2nd Derivatives,2105.714184,2174.386187


The bond_pv function returns the necessary inputs to calculate the next guess to solve for the yield to maturity:

*  The market price of the bond = the present value of cash flows based on spot rates (97.66).
*  The value of the cash flows based on the initial guess of 5.0%
*  The first derivative of the bond value function, evaluated at 5.0%.

The updated guess equals the previous guess plus the pricing error (bond price - previous guess) divided by the derivative.

In [ ]:
# pricing error
error=term_structure[0]-ytm_5[0]
# error is scaled by derivative value to find increment to guess
bond_derivative=term_structure[1]
delta_guess=error/bond_derivative
guess_new=guess+delta_guess
#revised guess is lower
guess_new

np.float64(0.04357848973553153)

In [ ]:
# results for yield to maturity with revised guess
guess=guess_new
ytm_new=bond_pv(guess,data_dict=data_dict)

# create DataFrame of results
results_dict={'5% Yield To Maturity':ytm_new,'Spot Rates':term_structure}
display(pd.DataFrame(results_dict,index=['Values','1st Derivatives','2nd Derivatives']))

,5% Yield To Maturity,Spot Rates
Values,97.612039,97.657391
1st Derivatives,-445.563702,-445.720259
2nd Derivatives,2173.745903,2174.386187


In [ ]:
# The revised yield to maturity guess results in a pricing error of less than 0.05%.
# If error is judged as acceptable, yield to maturity equals 4.35%
error=term_structure[0]-ytm_new[0]
error_pcnt=error/term_structure[0]*100
print(error_pcnt)

0.04644000261129602


## <span style="text-align: left; color:green; font-family: 'Franklin Gothic Medium', sans-serif; margin-top: 1.0em; margin-bottom: 0em; font-style: italic;">Chapter Exercise</span>


<div style="background-color:LightGray;
    border-left: 12px solid green;
    font-family: 'Garamond', serif;
    font-size: 17px;
    line-height: 1.5;
    padding: 15px">
<br>

Calculate the yield to maturity for the following bonds:


|Maturity|Coupon|Clean Price|
|-------|-------|-----------|
|&nbsp;&nbsp;&nbsp;June 30$^{st}$ 2035|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.375|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\$100|
|&nbsp;&nbsp;&nbsp;October 31$^{th}$ 2050|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.125|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\$95|
|&nbsp;&nbsp;&nbsp;January 31$^{th}$ 2040|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1.500|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\$65|

Assume that the settlement date is January 31$^{st}$ 2026
</div>

<div style="
    border-left: 12px solid green;
    line-height: 1.5;
    padding: 15px">
<br>


1. **Import Required Modules:**  
   * Built-in modules  
   * Pandas (install if necessary)  
   * Custom modules:  
     * <font color='green'>accrued_interest</font>  
     * <font color='green'>create_payoff_matrix</font>  
     * <font color='green'>bond_pay_data</font>  
2. **Add user defined functions:**
     * <font color='green'>bond_pv</font>  
     * <font color='green'>single_newton_raphson</font>  
     * <font color='green'>calc_ytm</font>  

3. **Create DataFrame for Bond Data:**  
   * Create dictionary for bond data
   * Convert dictionary to DataFrame with maturity dates as index  

4. **Perform Calculations:**  
   * Add accrued interest to DataFrame
   * Call calc_ytm

</div>

Required modules imported for January 31$^{st}$ 2030 Application
User defined function <font color='green'>bond_pv</font> added for January 31$^{st}$ 2030 Application
User defined functions <font color='green'>single_newton_raphson</font> and <font color='green'>calc_ytm</font> are added below.

In [ ]:
def single_newton_raphson(target_value, function=None, data=None, num_attempts=50,
                          guess=0.0001, tolerance=1e-5):
    """
    Finds a root of a function using the Newton-Raphson method.

    Args:
        target_value (float): The target value we want the function to return.
        function (callable): A function that takes a guess and optional data,
                             and returns a tuple (value, derivative).
        data (any, optional): Additional data to pass to the function.
        num_attempts (int): Maximum number of iterations.
        guess (float): Initial guess for the root.
        tolerance (float): The desired precision of the result.

    Returns:
        tuple: (number_of_guesses, final_guess, final_value)
        str: An error message if no solution is found.
    """
    if function is None:
      raise TypeError("function must be specified")
    import numpy as np
    num_guesses = 0

    for num_guesses in range(num_attempts):
       #Ignore second derivative
       value, derivative,*other = function(guess, data)

        # Calculate how far off we are from the target
       error = value - target_value

        # Check for convergence
       if abs(error) <= tolerance:
            return num_guesses, guess, value

        # Avoid division by zero
       if np.isclose(derivative, 0):
            print("No solution found: derivative is zero.")
            return num_guesses+1, guess, value

        # Newton-Raphson update step
        # We use value - target_value (the error) as f(x)
        # The standard formula is x_n+1 = x_n - f(x_n)/f'(x_n)
       guess = guess - error / derivative
    print("No solution found after {} attempts.".format(num_attempts))
    return num_guesses+1, guess, value

In [ ]:
def calc_ytm(price,maturity,coupon,guess=0.01,settlement=None,num_attempts=50,freq=2):
  """
  Calculates the Yield to Maturity (YTM) for a bond.

  YTM is the total annualized return an investor can expect if they hold the bond
  until it matures. This function finds the discount rate that equates the
  present value of the bond's future cash flows to its current market price
  using the Newton-Raphson numerical method.

  Args:
    price (float): The current market price of the bond.
    maturity (date): The bond's maturity date.
    coupon (float): The annual coupon rate of the bond (e.g., 0.05 for 5%).
    guess (float, optional): An initial guess for the YTM. Defaults to 0.01 (1%).
                           A good guess can speed up the calculation.
    settlement (date, optional): The date the bond is purchased.
                                 Defaults to the current date.

  Returns:
    float: The calculated Yield to Maturity (YTM) of the bond.

  Raises:
      TypeError: If 'price' is not a number
  """
  from datetime import date
  import numpy as np

  # If the initial guess is not a number (NaN), reset it to a default value.
  if np.isnan(guess):
    guess=max(0.01,coupon/100)

  # Validated price is number
  if np.isnan(price):
    raise TypeError('Price must be a number')

  # If no settlement date is provided, use today's date.
  if settlement is None:
    settlement=date.today()
  # Call bond_pay_data() to get the schedule of all future payments
  pay_data=bond_pay_data(maturity,coupon,settlement=settlement,freq=freq)

  # Create a dictionary to pass necessary data to the bond_pv function
  data_dict={'settlement':settlement,'pay_data':pay_data}

  #Newton-Raphson technique passes data to bond_pv to update estimates
  number_iterations,final_result,final_value=single_newton_raphson(price,
                                                                  function=bond_pv,
                                                                  data=data_dict,
                                                                  guess=guess)

  #Failure to converge causes alert, but final shows value at maximum tries
  return final_result

3. **Create DataFrame for Bond Data:**  
   * Create dictionary for bond data
   * Convert dictionary to DataFrame (<font color='green'>bond3_data</font>) with maturity dates as index  


In [ ]:
# Define bond data as a dictionary
data = {
    'Maturity': [date(2035,6,30), date(2050,10,31), date(2040,1,31)],
    'Coupon': [4.375, 4.125, 1.500],
    'Clean Price': [100, 95, 65]
}

# Create DataFrame with maturity as index
bond3_data = pd.DataFrame(data)
bond3_data=bond3_data.set_index('Maturity')
bond3_data

,Coupon,Clean Price
Maturity,,
2035-06-30,4.375,100
2050-10-31,4.125,95
2040-01-31,1.500,65


4. **Perform Calculations:**  
   * Add accrued interest to DataFrame
   * Call <font color='green'>calc_ytm</font>

In [ ]:
# settlement date is Jan 31, 2026
settlement=date(2026,1,31)

In [ ]:
# apply the accrued_interest function to the DataFrame
bond3_data['Accrued'] = bond_data.apply(
    lambda x: accrued_interest(
        maturity=x.name, # Assuming index is maturity
        coupon=x['Coupon'],
        settlement=settlement,
        freq=2,
        day_type='Actual/Actual'
    ), axis=1
)
bond3_data

,Coupon,Clean Price,Accrued
Maturity,,,
2035-06-30,4.375,100,0.374655
2050-10-31,4.125,95,1.053915
2040-01-31,1.500,65,0.000000


In [ ]:
# apply calc_ytm function to the DataFrame

bond3_data['Yield To Maturity'] = bond3_data.apply(
    lambda x: calc_ytm(
        maturity=x.name, # assuming index is maturity
# convert clean price to dirty by adding accrued interest
        price=(x['Clean Price']+x['Accrued']),
        coupon=x['Coupon'],
        guess=x['Coupon']/100, # convert coupon to decimal
        settlement=settlement,
        freq=2
    ), axis=1
)
bond3_data

,Coupon,Clean Price,Accrued,Yield To Maturity
Maturity,,,,
2035-06-30,4.375,100,0.374655,0.043268
2050-10-31,4.125,95,1.053915,0.044105
2040-01-31,1.500,65,0.000000,0.049467
